libr/io/map.c

```diff
@@ -1,372 +1,265 @@
-/* radare - LGPL - Copyright 2008-2016 - pancake */
+/* radare2 - LGPL - Copyright 2017 - condret */
 
-#include <stdio.h>
-#include <string.h>
-#include <stdlib.h>
 #include <r_io.h>
 #include <r_util.h>
-#include <r_list.h>
+#include <sdb.h>
+#include <stdlib.h>
 
-R_API int r_io_map_count (RIO *io) {
-	return ls_length (io->maps);
-}
+#define END_OF_MAP_IDS  0xffffffff
 
-R_API RIOMap * r_io_map_new(RIO *io, int fd, int flags, ut64 delta, ut64 addr, ut64 size) {
-	RIOMap *map = R_NEW0 (RIOMap);
-	if (!map || ((UT64_MAX - size) < addr)) { //prevent interger-overflow
+R_API RIOMap* r_io_map_new(RIO* io, int fd, int flags, ut64 delta, ut64 addr, ut64 size) {
+	RIOMap* map = NULL;
+	if (!size || !io || !io->maps || ((UT64_MAX - size + 1) < addr) || !io->map_ids) {
+		return NULL;
+	}
+	map = R_NEW0 (RIOMap);
+	if (!map || !io->map_ids || !r_id_pool_grab_id (io->map_ids, &map->id)) {
 		free (map);
 		return NULL;
 	}
 	map->fd = fd;
+	map->from = addr;
+	// RIOMap describes an interval of addresses (map->from; map->to)
+	map->to = addr + size - 1;
 	map->flags = flags;
 	map->delta = delta;
-	map->from = addr;
-	map->to = addr + size;
+	// new map lives on the top, being top the list's tail
 	ls_append (io->maps, map);
 	return map;
 }
 
-R_API void r_io_map_init(RIO *io) {
-	io->maps = ls_new ();
+static void _map_free(void* p) {
+	RIOMap* map = (RIOMap*) p;
+	if (map) {
+		free (map->name);
+		free (map);
+	}
 }
 
-R_API int r_io_map_sort(void *_a, void *_b) {
-	RIOMap *a = _a, *b = _b;
-	if (a->from == b->from) {
-		ut64 a_sz = a->to - a->from;
-		ut64 b_sz = b->to - b->from;
-		return a_sz < b_sz;
+R_API void r_io_map_init(RIO* io) {
+	if (io && !io->maps) {
+		io->maps = ls_newf ((SdbListFree)_map_free);
+		if (io->map_ids) {
+			r_id_pool_free (io->map_ids);
+		}
+		io->map_ids = r_id_pool_new (1, END_OF_MAP_IDS);
 	}
-	return a->from < b->from;
 }
 
-R_API int r_io_map_write_update(RIO *io, int fd, ut64 addr, ut64 len) {
-	int res = false;
-	RIOMap *map = NULL;
-	SdbListIter *iter;
-	ls_foreach_prev (io->maps, iter, map) {
-		if (map->fd == fd) break;
-		map = NULL;
+// check if a map with exact the same properties exists
+R_API bool r_io_map_exists(RIO* io, RIOMap* map) {
+	SdbListIter* iter;
+	RIOMap* m;
+	if (!io || !io->maps || !map) {
+		return false;
 	}
-	if (map && map->to < addr+len) {
-		res = true;
-		map->to = addr+len;
+	ls_foreach (io->maps, iter, m) {
+		if (!memcmp (m, map, sizeof (RIOMap))) {
```

Comment:
hm? the idea for this was, checking if a map belongs to an RIO-instance, if you only have the data, and not the pointer, can remove dat later if you want